File: app/src/main/java/com/absinthe/anywhere_/model/AnywhereDao.java
package com.absinthe.anywhere_.model;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Update;

import java.util.List;

@Dao
public interface AnywhereDao {

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insert(AnywhereEntity ae);

    @Update
    void update(AnywhereEntity ae);

    @Query("DELETE FROM anywhere_table")
    void deleteAll();

    @Delete
    void delete(AnywhereEntity ae);

    @Query("SELECT * from anywhere_table ORDER BY time_stamp DESC")
    LiveData<List<AnywhereEntity>> getAllAnywhereEntitiesOrderByTimeDesc();

    @Query("SELECT * from anywhere_table ORDER BY time_stamp ASC")
    LiveData<List<AnywhereEntity>> getAllAnywhereEntitiesOrderByTimeAsc();

    @Query("SELECT * from anywhere_table ORDER BY app_name DESC")
    LiveData<List<AnywhereEntity>> getAllAnywhereEntitiesOrderByNameDesc();

    @Query("SELECT * from anywhere_table ORDER BY app_name ASC")
    LiveData<List<AnywhereEntity>> getAllAnywhereEntitiesOrderByNameAsc();

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertPage(PageEntity pe);

    @Update
    void updatePage(PageEntity pe);

    @Delete
    void deletePage(PageEntity pe);

    @Query("SELECT * from page_table ORDER BY priority ASC")
    LiveData<List<PageEntity>> getAllPageEntities();
}

File: app/src/main/java/com/absinthe/anywhere_/database/AnywhereDao.kt
package com.absinthe.anywhere_.database

import android.database.Cursor
import androidx.lifecycle.LiveData
import androidx.room.*
import com.absinthe.anywhere_.model.database.AnywhereEntity
import com.absinthe.anywhere_.model.database.PageEntity

@Dao
interface AnywhereDao {
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insert(ae: AnywhereEntity)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insert(list: List<AnywhereEntity>)

    @Update
    suspend fun update(ae: AnywhereEntity)

    @Update
    suspend fun update(list: List<AnywhereEntity>)

    @Query("DELETE FROM anywhere_table")
    suspend fun deleteAll()

    @Delete
    suspend fun delete(ae: AnywhereEntity)

    @Delete
    suspend fun delete(list: List<AnywhereEntity>)

    @get:Query("SELECT * from anywhere_table ORDER BY time_stamp DESC")
    val allAnywhereEntitiesOrderByTimeDesc: LiveData<List<AnywhereEntity>>

    @get:Query("SELECT * from anywhere_table ORDER BY time_stamp ASC")
    val allAnywhereEntitiesOrderByTimeAsc: LiveData<List<AnywhereEntity>>

    @get:Query("SELECT * from anywhere_table ORDER BY app_name DESC")
    val allAnywhereEntitiesOrderByNameDesc: LiveData<List<AnywhereEntity>>

    @get:Query("SELECT * from anywhere_table ORDER BY app_name ASC")
    val allAnywhereEntitiesOrderByNameAsc: LiveData<List<AnywhereEntity>>

    @Query("SELECT * from anywhere_table WHERE _id LIKE :id")
    fun getEntityById(id: String): AnywhereEntity?

    @Query("SELECT * FROM anywhere_table")
    fun selectAll(): Cursor?

    @Query("SELECT * FROM anywhere_table WHERE _id LIKE :id")
    fun selectById(id: Long): Cursor?

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertPage(pe: PageEntity)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertPage(pageList: List<PageEntity>)

    @Update
    suspend fun updatePage(pe: PageEntity)

    @Delete
    suspend fun deletePage(pe: PageEntity)

    @get:Query("SELECT * from page_table ORDER BY priority ASC")
    val allPageEntities: LiveData<List<PageEntity>>

    @Query("SELECT * from page_table WHERE title LIKE :title")
    fun getPageEntityByTitle(title: String): PageEntity?
}